price-estimator/src/solver_rounding_buffer.rs

```diff
@@ -0,0 +1,162 @@
+    for o in orders.iter_mut() {
+        // Compute rounding buffer.
+        let (t_s, t_b) = (TokenId(o.sell_token), TokenId(o.buy_token));
+        let estimated_buy_token_price = match tokens.get(&t_b) {
+            Some(TokenInfo { external_price, .. }) if *external_price > 0 => *external_price as f64,
+            _ => continue,
+        };
+        let estimated_sell_token_price = match tokens.get(&t_s) {
+            Some(TokenInfo { external_price, .. }) if *external_price > 0 => *external_price as f64,
+            _ => continue,
+        };
+        let estimated_xrate = estimated_buy_token_price / estimated_sell_token_price;
+        let rounding_buffer = max_rounding_amount(estimated_buy_token_price, fee_token_price)
+            * estimated_xrate
+            * PRICE_ESTIMATION_ERROR.powi(2);
+        let rounding_buffer = rounding_buffer.ceil() as u128;
+
+        // Update rounding buffer for account balances;
+        let entry = account_balance_buffers
+            .entry((o.account_id, t_s))
+            .or_default();
+        *entry = (*entry).max(rounding_buffer);
+
+        // Reduce order sell amount.
+        o.denominator = o.denominator.saturating_sub(rounding_buffer);
+        o.remaining_sell_amount = o.remaining_sell_amount.saturating_sub(rounding_buffer);
+    }
+
+    // Reduce account balances.
+    for ((address, token_id), rounding_buffer) in account_balance_buffers {
+        if let Some(balance) = account_state.0.get_mut(&(address, token_id.0)) {
+            *balance = balance.saturating_sub(U256::from(rounding_buffer));
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ethcontract::{Address, U256};
+
+    fn address(n: u64) -> Address {
+        Address::from_low_u64_le(n)
+    }
+
+    fn account(address_: u64, token: u16, balance: u128) -> ((Address, u16), U256) {
+        ((address(address_), token), U256::from(balance))
+    }
+
+    fn token_info(external_price: u128) -> TokenInfo {
+        TokenInfo {
+            alias: None,
+            decimals: None,
+            external_price,
+        }
+    }
+
+    fn order(
+        id: u16,
+        address_: u64,
+        buy_token: u16,
+        sell_token: u16,
+        numerator: u128,
+        denominator: u128,
+    ) -> Order {
+        Order {
+            id,
+            account_id: address(address_),
+            buy_token,
+            sell_token,
+            numerator,
+            denominator,
+            remaining_sell_amount: denominator,
+            valid_from: 0,
+            valid_until: 0,
+        }
+    }
+
+    #[test]
+    fn apply_rounding_buffer_ok() {
+        let mut tokens = Tokens::new();
+        tokens.insert(TokenId(0), token_info(1));
+        tokens.insert(TokenId(1), token_info(2));
+        tokens.insert(TokenId(2), token_info(10));
+
+        let accounts = vec![
+            account(0, 0, 100_000_000_000_000_000),
```

Comment:
I assume you don't think the test macros are worth the effort here? They would allow reading the test setup in a non IDE without jumping back and forth to the account/order function to remember in which order the 3/6 numbers are passed into the named struct.

Reply:
For readability I want these account and order definitions to fit in one line (or maybe 2). With the full structs it becomes too large. In this case since the functions are right above the test and there are only two functions I think it's fine and when reading you'll remember the parameters after looking them up once.
If it was more functions and more tests maybe macro would be worth it but not at this point.